test/session.test.js

```diff
@@ -0,0 +1,160 @@
+    this.timeout(Factory.TestTimeout);
+    this.expectedItemCount = BASE_ITEM_COUNT;
+    this.application = await Factory.createInitAppWithRandNamespace();
+    this.email = Uuid.GenerateUuidSynchronously();
+    this.password = Uuid.GenerateUuidSynchronously();
+    await Factory.registerUserToApplication({
+      application: this.application,
+      email: this.email,
+      password: this.password
+    });
+  });
+
+  afterEach(async function () {
+    expect(this.application.syncService.isOutOfSync()).to.equal(false);
+    const items = this.application.itemManager.items;
+    expect(items.length).to.equal(this.expectedItemCount);
+    const rawPayloads = await this.application.storageService.getAllRawPayloads();
+    expect(rawPayloads.length).to.equal(this.expectedItemCount);
+    await this.application.deinit();
+  });
+
+  function getDelayBeforeNextRequest(currentSession) {
+    const tokenExpireAt = currentSession.expireAt;
+    const timeRemaining = Math.ceil(tokenExpireAt - Date.now() / 1000);
+    /* 
+      If the token has not expired yet, we will return the remaining time, plus one second.
+      Else, there's no need to add a delay.
+    */
+    return timeRemaining > 0 ? timeRemaining + 1 : 0;
+  }
+
+  async function getSessionFromStorage(application) {
+    return application.storageService.getValue(StorageKey.Session);
+  }
+
+  it('should fail when a sync request is perfomed with an expired access token', async function () {
+    const currentSession = this.application.apiService.session;
+
+    const delayBeforeNextRequest = getDelayBeforeNextRequest(currentSession);
+    await Factory.sleep(delayBeforeNextRequest);
+
+    const response = await this.application.apiService.sync([]);
+
+    expect(response).to.have.property('status');
+    expect(response.status).to.equal(498);
+
+    expect(response).to.have.property('error');
+
+    expect(response.error).to.have.property('tag');
+    expect(response.error.tag).to.equal('expired-access-token');
+
+    expect(response.error).to.have.property('message');
+    expect(response.error.message).to.equal('The provided access token has expired.');
+  }).timeout(10000);
+
+  it('should return the new session in the response when refreshed', async function () {
+    const response = await this.application.apiService.refreshSession();
+
+    expect(response).to.have.property('status');
+    expect(response.status).to.equal(200);
+
+    expect(response).to.have.property('token');
+    expect(response.token).to.be.a('string');
+    expect(response.token).to.not.be.empty;
+
+    expect(response).to.have.property('session');
+
+    expect(response.session).to.have.property('expire_at');
+    expect(response.session.expire_at).to.be.a('number');
+
+    expect(response.session).to.have.property('refresh_token');
+    expect(response.session.refresh_token).to.not.be.empty;
+  }).timeout(10000);
+
+  it('should be refreshed if access token is expired', async function () {
```

Comment:
This is a good test 👍 Another thing we should probably test for is expecting that the new token expireAt > Date.now().